src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -658,6 +658,25 @@ public ProjectState UpdateAdditionalDocument(TextDocumentState newDocument, bool
                 latestDocumentTopLevelChangeVersion: dependentSemanticVersion);
         }
 
+        public ProjectState UpdateDocumentsOrder(ImmutableArray<DocumentId> documentIds)
+        {
+            if (documentIds.Length != _documentIds.Count)
+            {
+                throw new ArgumentException(WorkspacesResources.The_count_of_the_specified_document_ids_do_not_match_the_count_of_the_document_ids_in_the_project, nameof(documentIds));
+            }
+
+            foreach(var documentId in documentIds)
+            {
+                if (!this.ContainsDocument(documentId))
+                {
+                    throw new InvalidOperationException(String.Format(WorkspacesResources.Document_id_0_from_the_specified_document_ids_does_not_exist_within_the_project, documentId.Id));
+                }
+            }
+
+            return this.With(
+                documentIds: documentIds.ToImmutableList());
```

Comment:
Given we hold an ImmutableList under the covers, should this API just take that? Or at least not take an ImmutableArray? My worry if you have a project with a thousand files in it, this is a bit allocation heavy vs. our ability to reuse the underlying ImmutableList that we had prior.

Reply:
Maybe that makes sense. I can change it to ImmutableList.